In [11]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn import tree
from sklearn import preprocessing
from sklearn.metrics import accuracy_score, log_loss
from sklearn import metrics
import graphviz
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
import matplotlib.pyplot as plt
from sklearn.metrics import classification_report, confusion_matrix

In [9]:
data = pd.read_csv("Titanic.csv")
data = data[data['age'].notna()]

le = preprocessing.LabelEncoder()

data['pclass'] = le.fit_transform(data['pclass'])

data['sex'] = le.fit_transform(data['sex'])

features = ['pclass', 'sex', 'age', 'sibsp']

y = data['survived']
x = data[features]

x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2)

In [14]:
clf = MLPClassifier(solver='adam', alpha=0.001, hidden_layer_sizes=(50, 20), 
                    random_state=1, activation='logistic', learning_rate_init=0.001) #logistic activation (i.e sigmoid)

clf.fit(x_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='logistic', alpha=0.001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(50, 20), learning_rate='constant',
              learning_rate_init=0.001, max_fun=15000, max_iter=200,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=1, shuffle=True, solver='adam',
              tol=0.0001, validation_fraction=0.1, verbose=False,
              warm_start=False)

In [15]:
y_pred = clf.predict(x_test)
#print(y_pred) #0 for survival, 1 for death

print(accuracy_score(y_test, y_pred))

0.8476190476190476


In [16]:

clf = MLPClassifier(solver='adam', alpha=0.001, hidden_layer_sizes=(15, 8), random_state=1, activation='logistic', learning_rate_init=0.001) #logistic activation

clf.fit(x_train, y_train)


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='logistic', alpha=0.001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(15, 8), learning_rate='constant',
              learning_rate_init=0.001, max_fun=15000, max_iter=200,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=1, shuffle=True, solver='adam',
              tol=0.0001, validation_fraction=0.1, verbose=False,
              warm_start=False)

In [17]:

y_pred = clf.predict(x_test)
#print(y_pred) #0 for survival, 1 for death
print(accuracy_score(y_test, y_pred))


0.819047619047619


In [18]:
clf = MLPClassifier(solver='adam', alpha=0.001, hidden_layer_sizes=(50, 20), random_state=1, activation='relu', learning_rate_init=0.001) 

clf.fit(x_train, y_train)


MLPClassifier(activation='relu', alpha=0.001, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(50, 20), learning_rate='constant',
              learning_rate_init=0.001, max_fun=15000, max_iter=200,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=1, shuffle=True, solver='adam',
              tol=0.0001, validation_fraction=0.1, verbose=False,
              warm_start=False)

In [19]:
y_pred = clf.predict(x_test)
#print(y_pred) #0 for survival, 1 for death
print(accuracy_score(y_test, y_pred))


0.8380952380952381


In [20]:
clf = MLPClassifier(solver='adam', alpha=0.001, hidden_layer_sizes=(20, 5), random_state=1, activation='logistic', learning_rate_init=0.01, max_iter=400) #logistic activation (i.e sigmoid)

clf.fit(x_train, y_train)


y_pred = clf.predict(x_test)
#print(y_pred) #0 for survival, 1 for death

print(accuracy_score(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))

0.8428571428571429
[[112  14]
 [ 19  65]]


ANN Method:

[[112, 14]

[19, 65]]

True Negatives: 112 
False Negatives: 14

False Positives: 19
True Positives: 65


Decision Tree Method:

[[116 14]

[[ 28 52]]

True Negatives: 116, False Negatives: 14

False Positives: 28, True Positives: 52